<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [Mono-devel-list] Re: [PATCH] Async bugs with Sockets
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:mono-devel-list%40lists.ximian.com?Subject=%5BMono-devel-list%5D%20Re%3A%20%5BPATCH%5D%20Async%20bugs%20with%20Sockets&In-Reply-To=Pine.LNX.4.55.0312271645321.517%40danga.com">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="003380.html">
   <LINK REL="Next"  HREF="003398.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[Mono-devel-list] Re: [PATCH] Async bugs with Sockets</H1>
    <B>Brad Fitzpatrick</B> 
    <A HREF="mailto:mono-devel-list%40lists.ximian.com?Subject=%5BMono-devel-list%5D%20Re%3A%20%5BPATCH%5D%20Async%20bugs%20with%20Sockets&In-Reply-To=Pine.LNX.4.55.0312271645321.517%40danga.com"
       TITLE="[Mono-devel-list] Re: [PATCH] Async bugs with Sockets">brad at danga.com
       </A><BR>
    <I>Sun Dec 28 20:26:11 EST 2003</I>
    <P><UL>
        <LI>Previous message: <A HREF="003380.html">[Mono-devel-list] Async bugs with Sockets
</A></li>
        <LI>Next message: <A HREF="003398.html">[Mono-devel-list] Re: [PATCH] Async bugs with Sockets
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#3384">[ date ]</a>
              <a href="thread.html#3384">[ thread ]</a>
              <a href="subject.html#3384">[ subject ]</a>
              <a href="author.html#3384">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Bugzilla's still down, so:

Patch:
  <A HREF="http://www.bradfitz.com/share/mono/Socket-async-patch.txt">http://www.bradfitz.com/share/mono/Socket-async-patch.txt</A>

Testcase:
  <A HREF="http://www.bradfitz.com/share/mono/AsyncSocketTest.cs">http://www.bradfitz.com/share/mono/AsyncSocketTest.cs</A>

Failed before, now passes.  matches Microsoft.NET behavior.

I only protected Connect/Send/SendTo, because that's all I've seen fail,
but I'm sure the others do too.  Before I do more, I'd like to know if I'm
even on the right track.

Please critique without holding back.  I want to see this fixed properly
moreso than I want to see my code in the tree, so if this is all wrong
let me know.

Attached the files too, not knowing proper etiquette around here.  (I'm
pretty sure I've read all available docs on the mono website, but I don't
remember anything on this issue)

- Brad



On Sat, 27 Dec 2003, Brad Fitzpatrick wrote:

&gt;<i> I'm writing a reverse-proxy load balancer and so far it's been
</I>&gt;<i> amazingly easy with hardly any code, but I've been bitten by these two
</I>&gt;<i> bugs:
</I>&gt;<i>
</I>&gt;<i> (Sorry, I can't reach bugzilla.ximian.com for the past couple days)
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> Bug #1:
</I>&gt;<i>
</I>&gt;<i> None of the async methods in Socket.cs catch exceptions in the Worker
</I>&gt;<i> thread and return them to the caller thread during the End* methods.
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> Consider this simple example:
</I>&gt;<i>
</I>&gt;<i>         public static void Main ()
</I>&gt;<i>         {
</I>&gt;<i>             IPAddress ipOne = IPAddress.Parse(&quot;10.5.2.1&quot;);   // something bogus
</I>&gt;<i>             IPEndPoint ipEP = new IPEndPoint(ipOne, 23483);  // something bogus
</I>&gt;<i>             Socket pSock = new Socket(ipEP.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
</I>&gt;<i>
</I>&gt;<i>             IAsyncResult ar = pSock.BeginConnect(ipEP, null, null);
</I>&gt;<i>
</I>&gt;<i>             try {
</I>&gt;<i>                 pSock.EndConnect(ar);  // should raise an exception because connect was bogus
</I>&gt;<i>             } catch {
</I>&gt;<i>                 Console.WriteLine(&quot;connect error&quot;);
</I>&gt;<i>                 return;
</I>&gt;<i>             }
</I>&gt;<i>         }
</I>&gt;<i>
</I>&gt;<i> With Mono:
</I>&gt;<i>
</I>&gt;<i> $ ./AsyncErrorDemo2.exe
</I>&gt;<i>
</I>&gt;<i> Unhandled Exception: System.Net.Sockets.SocketException: Connection refused
</I>&gt;<i> in (unmanaged) /usr/local/lib/libmono.so.0(mono_raise_exception+0x1f) [0x4008f15f]
</I>&gt;<i> in (unmanaged) /usr/local/lib/libmono.so.0 [0x400afac5]
</I>&gt;<i> in &lt;0x00028&gt; (wrapper managed-to-native) System.Net.Sockets.Socket:Connect_internal (intptr,System.Net.SocketAddress)
</I>&gt;<i> in &lt;0x0004e&gt; System.Net.Sockets.Socket:Connect (System.Net.EndPoint)
</I>&gt;<i> in &lt;0x0006c&gt; .Worker:Connect ()
</I>&gt;<i> in &lt;0x00044&gt; (wrapper delegate-invoke) System.MulticastDelegate:invoke_void ()
</I>&gt;<i>
</I>&gt;<i> (Couldn't catch it)
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> With .NET:
</I>&gt;<i>
</I>&gt;<i> $ ./AsyncErrorDemo2.exe
</I>&gt;<i> connect error
</I>&gt;<i>
</I>&gt;<i> (correct)
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> Proposed fix:
</I>&gt;<i>
</I>&gt;<i> In mcs-0.29/class/System/System.Net.Sockets/Socket.cs, add a member to
</I>&gt;<i> SocketAsyncResult to hold the exception.  Then all the methods in the
</I>&gt;<i> Worker class wrap their synchronous methods with try/catch, saving the
</I>&gt;<i> exception in their SocketAsyncResult result.  Then the Socket.End* methods
</I>&gt;<i> rethrow that exception if present.
</I>&gt;<i>
</I>&gt;<i> How do I go about submitting such a fix?  Are there subsystem maintainers,
</I>&gt;<i> or do I just submit a patch to bugzilla/this list?  (I'm a mono hacker newbie)
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> Bug #2:
</I>&gt;<i>
</I>&gt;<i> Socket.BeginConnect() is documented on MSDN to throw an exception
</I>&gt;<i> synchronously if the socket isn't connected:
</I>&gt;<i>
</I>&gt;<i>    &quot;BeginSend will throw an exception if you do not first call Accept,
</I>&gt;<i>     BeginAccept, Connect, or BeginConnect.&quot;
</I>&gt;<i>
</I>&gt;<i> And indeed:
</I>&gt;<i>
</I>&gt;<i>         public static void Main ()
</I>&gt;<i>         {
</I>&gt;<i>             IAsyncResult ar;
</I>&gt;<i>
</I>&gt;<i>             Console.WriteLine(&quot;demo start&quot;);
</I>&gt;<i>
</I>&gt;<i>             IPAddress ipOne = IPAddress.Parse(&quot;10.0.0.1&quot;);
</I>&gt;<i>             IPEndPoint ipEP = new IPEndPoint(ipOne, 80);
</I>&gt;<i>             Socket pSock = new Socket(ipEP.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
</I>&gt;<i>
</I>&gt;<i>             byte[] buf = { 1, 1, 1 };
</I>&gt;<i>             ar = pSock.BeginSend(buf, 0, buf.Length, SocketFlags.None, null, null);
</I>&gt;<i>
</I>&gt;<i>             int rv = -1;
</I>&gt;<i>             try {
</I>&gt;<i>                 Console.WriteLine(&quot;demo middle (before EndSend)&quot;);
</I>&gt;<i>                 rv = pSock.EndSend(ar);
</I>&gt;<i>                 Console.WriteLine(&quot;EndSend() = {0}&quot;, rv);
</I>&gt;<i>             } catch (Exception e) {
</I>&gt;<i>                 Console.WriteLine(&quot;exception: &quot; + e.ToString());
</I>&gt;<i>             }
</I>&gt;<i>
</I>&gt;<i>             Console.WriteLine(&quot;demo end&quot;);
</I>&gt;<i>
</I>&gt;<i>         }
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> Mono: (wrong)
</I>&gt;<i>
</I>&gt;<i> demo start
</I>&gt;<i>
</I>&gt;<i> Unhandled Exception: System.Net.Sockets.SocketException: The socket has been shut down
</I>&gt;<i> in &lt;0x00102&gt; System.Net.Sockets.Socket:Send (byte[],int,int,System.Net.Sockets.SocketFlags)
</I>&gt;<i> in &lt;0x00056&gt; .Worker:Send ()
</I>&gt;<i> in &lt;0x00044&gt; (wrapper delegate-invoke) System.MulticastDelegate:invoke_void ()
</I>&gt;<i>
</I>&gt;<i> demo middle (before EndSend)
</I>&gt;<i>
</I>&gt;<i> &lt;... hangs, because worker thread is dead ...&gt;
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> .NET:  (Correct)
</I>&gt;<i>
</I>&gt;<i> $ ./AsyncErrorDemo2.exe
</I>&gt;<i> demo start
</I>&gt;<i>
</I>&gt;<i> Unhandled Exception: System.Net.Sockets.SocketException: A request to send or receive data was disallowed because the socket is not connected and (when sending on a datagram socket using a sendto call) no address was supplied
</I>&gt;<i>    at System.Net.Sockets.Socket.BeginSend(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags, AsyncCallback callback, Object state)
</I>&gt;<i>    at Danga.AsyncErrorDemo2.Main()
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> Also seems easy to fix.
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> So, what's the best way to get fixes for these in?  I'm willing to do
</I>&gt;<i> any legwork, unit tests, etc.  Just tell me what hoops to jump through
</I>&gt;<i> and who to talk to.  I'm loving C# and Mono and want to help.
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> --
</I>&gt;<i> Brad Fitzpatrick
</I>&gt;<i> <A HREF="http://www.danga.com/">http://www.danga.com/</A>
</I>&gt;<i> <A HREF="http://lists.ximian.com/mailman/listinfo/mono-devel-list">brad at danga.com</A>
</I>&gt;<i>
</I>&gt;<i> _______________________________________________
</I>&gt;<i> Mono-devel-list mailing list
</I>&gt;<i> <A HREF="http://lists.ximian.com/mailman/listinfo/mono-devel-list">Mono-devel-list at lists.ximian.com</A>
</I>&gt;<i> <A HREF="http://lists.ximian.com/mailman/listinfo/mono-devel-list">http://lists.ximian.com/mailman/listinfo/mono-devel-list</A>
</I>&gt;<i>
</I>&gt;<i>
</I>-------------- next part --------------
// System.Net.Sockets.TcpClientTest.cs

//

// Authors:

//    Brad Fitzpatrick (<A HREF="http://lists.ximian.com/mailman/listinfo/mono-devel-list">brad at danga.com</A>)

//

// (C) Copyright 2003 Brad Fitzpatrick

//



using System;

using System.Net;

using System.Net.Sockets;

using NUnit.Framework;



namespace MonoTests.System.Net.Sockets {



	/// &lt;summary&gt;

	/// Tests async behavior of System.Net.Sockets.Socket

	/// &lt;/summary&gt;

	[TestFixture]

	public class AsyncSocketTest {

		

		/// &lt;summary&gt;

		/// Tests whether exceptions are returned by the EndConnect

		/// &lt;/summary&gt;

		[Test]

		public void EndConnect ()

		{

		    IPAddress ipOne = IPAddress.Parse(&quot;127.1.2.3&quot;);   // something bogus

		    IPEndPoint ipEP = new IPEndPoint(ipOne, 23483);  // something bogus

		    Socket sock = new Socket(ipEP.AddressFamily, SocketType.Stream, ProtocolType.Tcp);

		    IAsyncResult ar = sock.BeginConnect(ipEP, null, null);

		    bool gotException = false;



		    try {

			sock.EndConnect(ar);  // should raise an exception because connect was bogus

		    } catch {

			gotException = true;

		    }



		    Assertion.AssertEquals(&quot;A01&quot;, gotException, true);

		}	

		

	}



}

-------------- next part --------------
An embedded and charset-unspecified text was scrubbed...
Name: Socket-async-patch.txt
Url: <A HREF="http://lists.ximian.com/pipermail/mono-devel-list/attachments/20031228/8554757a/attachment.txt">http://lists.ximian.com/pipermail/mono-devel-list/attachments/20031228/8554757a/attachment.txt</A> 
</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="003380.html">[Mono-devel-list] Async bugs with Sockets
</A></li>
	<LI>Next message: <A HREF="003398.html">[Mono-devel-list] Re: [PATCH] Async bugs with Sockets
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#3384">[ date ]</a>
              <a href="thread.html#3384">[ thread ]</a>
              <a href="subject.html#3384">[ subject ]</a>
              <a href="author.html#3384">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://lists.ximian.com/mailman/listinfo/mono-devel-list">More information about the Mono-devel-list
mailing list</a><br>
</body></html>
